docs/architecture-decision-records/002-component-constuctors.md

```diff
@@ -0,0 +1,144 @@
+
+## Positions
+
+<!--- What are the differing positions or proposals on this issue? -->
+
+1. Constructors should be the same or as similar as possible for consistency
+
+   By having a consistent style for constructors, users will know what to expect every time they use a class from the library. This makes the developer experience easier and faster as users do not have to check what is required for each class.
+
+2. Constrctors should be the best fit for each class
+
+   In the case of this library, this mainly means either missing out the id or changing the order of the id and props inputs. In scenarios where a sensible id can either be hardcoded or defaulted, this would prevent users from having to add ids everywhere, reducing the amount of code they have to write.
+
+## Decision
+
+<!-- What is the change that we're proposing and/or doing? -->
+
+Constructors should follow the following rules for consistency.
+
+1. Constructors should all accept a `scope` of type `GuStack` and an `id` of type `string`
+
+   :white_check_mark: Valid
+
+   ```ts
+   class MyConstruct {
+     constructor(scope: GuStack, id: string) {
+       ...
+     }
+   }
+   ```
+
+   :x: Invalid
+
+   ```ts
+   class MyConstruct {
+     constructor(scope: Stack, id: string) {
+       ...
+     }
+   }
+   ```
+
+2. They can also take a `props` object which should be correctly typed
+
+   :white_check_mark: Valid
+
+   ```ts
+   interface MyConstructProps {...}
+
+   class MyConstruct {
+     constructor(scope: GuStack, id: string, props: MyConstructProps) {
+       ...
+     }
+   }
+   ```
+
+   :x: Invalid
+
+   ```ts
+   class MyConstruct {
+     constructor(scope: GuStack, id: string, props: object) {
+       ...
+     }
+   }
+   ```
+
+3. Where all `props` are optional, the `props` object should be optional as a whole
+
+   :white_check_mark: Valid
+
+   ```ts
+   interface MyConstructProps {
+     prop1?: string;
+     prop2?: string
+   }
+
+   class MyConstruct {
+     constructor(scope: GuStack, id: string, props?: MyConstructProps) {
+       ...
+     }
+   }
+   ```
+
+   :x: Invalid
+
+   ```ts
+   interface MyConstructProps {
+     prop1?: string;
+     prop2?: string
+   }
+
+   class MyConstruct {
+     constructor(scope: GuStack, id: string, props: MyConstructProps) {
+       ...
+     }
+   }
+   ```
+
+4. Where `props` are optional, a default value for the `id` can be provided where appropriate
+
+   :white_check_mark: Valid
+
+   ```ts
+   interface MyConstructProps {
+     prop1?: string;
+     prop2?: string
+   }
+
+   class MyConstruct {
+     constructor(scope: GuStack, id: string = "MyConstruct", props?: MyConstructProps) {
```

Comment:
Opinions wanted on this pattern that I've been playing around with locally:

```ts
class Foo {
  constructor(scope: GuStack, id: string, props?: FooProps) {
    // do stuff
  }

  static withProps(scope: GuStack, props?: FooProps) {
    return new Foo(scope, "MyFoo", props)
  }
}

class Bar {
  private foo: Foo;

  constructor(props: FooProps) {
    this.foo = Foo.withProps(props);
  }
}
```

The idea here is to avoid needing to specify the ID when you want to provide props; this is the closest thing I can find to an overloaded constructor in TypeScript.

Reply:
Interesting. I'd certainly favour this over removing the ID from the constructor. Maybe another one for discussions?